src/systems/thruster/Thruster.cc

```diff
@@ -0,0 +1,258 @@
+
+  this->dataPtr->jointAxis =
+    _ecm.Component<ignition::gazebo::components::JointAxis>(jointEntity)
+      ->Data().Xyz();
+
+  std::string thrusterTopic = ignition::transport::TopicUtils::AsValidTopic(
+    "/model/" + ns + "/joint/" + jointName + "/cmd_pos");
+
+  this->dataPtr->node.Subscribe(
+    thrusterTopic,
+    &ThrusterPrivateData::OnCmdThrust,
+    this->dataPtr.get());
+
+  // Get link entity
+  this->dataPtr->linkEntity = model.LinkByName(_ecm, childLink->Data());
+
+  // Create an angular velocity component if one is not present.
+  if (!_ecm.Component<ignition::gazebo::components::AngularVelocity>(
+      this->dataPtr->linkEntity))
+  {
+    _ecm.CreateComponent(this->dataPtr->linkEntity,
+      ignition::gazebo::components::AngularVelocity());
+  }
+
+    // Create an angular velocity component if one is not present.
+  if (!_ecm.Component<ignition::gazebo::components::WorldAngularVelocity>(
+      this->dataPtr->linkEntity))
+  {
+    _ecm.CreateComponent(this->dataPtr->linkEntity,
+      ignition::gazebo::components::WorldAngularVelocity());
+  }
+
+  double p         =  0.1;
+  double i         =  0;
+  double d         =  0;
+  double iMax      =  1;
+  double iMin      = -1;
+  double cmdMax    = this->dataPtr->ThrustToAngularVec(this->dataPtr->cmdMax);
+  double cmdMin    = this->dataPtr->ThrustToAngularVec(this->dataPtr->cmdMin);
+  double cmdOffset =  0;
+
+  if (_sdf->HasElement("p_gain"))
+  {
+    p = _sdf->Get<double>("p_gain");
+  }
+  if (!_sdf->HasElement("i_gain"))
+  {
+    i = _sdf->Get<double>("i_gain");
+  }
+  if (!_sdf->HasElement("d_gain"))
+  {
+    d = _sdf->Get<double>("d_gain");
+  }
+
+  this->dataPtr->rpmController.Init(
+    p,
+    i,
+    d,
+    iMax,
+    iMin,
+    cmdMax,
+    cmdMin,
+    cmdOffset);
+}
+
+void ThrusterPrivateData::OnCmdThrust(const ignition::msgs::Double &_msg)
+{
+  std::lock_guard<std::mutex> lock(mtx);
+  this->thrust = ignition::math::clamp(ignition::math::fixnan(_msg.data()),
+    this->cmdMin, this->cmdMax);
+}
+
+double ThrusterPrivateData::ThrustToAngularVec(double _thrust)
+{
+  // Thrust is proprtional to the Rotation Rate squared
+  // See Thor I Fossen's  "Guidance and Control of ocean vehicles" p. 246
+  auto propAngularVelocity = sqrt(abs(
+    _thrust /
+      (this->fluidDensity
+      * this->thrustCoefficient * pow(this->propellerDiameter, 4))));
+
+  propAngularVelocity *= (_thrust > 0) ? 1: -1;
+
+  return propAngularVelocity;
+}
+
+void Thruster::PreUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (_info.paused)
+    return;
+
+  ignition::gazebo::Link link(this->dataPtr->linkEntity);
+
+  auto pose = worldPose(this->dataPtr->linkEntity, _ecm);
+
+  // TODO(arjo129): add logic for custom coordinate frame
+  auto unitVector = pose.Rot().RotateVector(
+    this->dataPtr->jointAxis.Normalize());
+
+  std::lock_guard<std::mutex> lock(this->dataPtr->mtx);
```

Comment:
Do you need to hold the lock until the end of the function? You can probably use a new scope `{` `}` and surround just the line to read `this->dataPtr->thrust`.